src/ML-POD/fitpod_command.cpp

```diff
@@ -0,0 +1,1617 @@
+
+  if (narg > 2)
+    coeff_file = std::string(arg[2]); // coefficient input file
+  else
+    coeff_file = "";
+
+  podptr = new MLPOD(lmp, pod_file, coeff_file);
+  read_data_files(data_file, podptr->pod.species);
+
+  if ((int) traindata.data_path.size() > 1)
+    allocate_memory(traindata);
+  else if ((int) testdata.data_path.size() > 1)
+    allocate_memory(testdata);
+
+  // get POD coefficients from an input file
+
+  if (coeff_file != "") podptr->podArrayCopy(desc.c, podptr->pod.coeff, podptr->pod.nd);
+
+  // compute POD coefficients using least-squares method
+
+  least_squares_fit(traindata);
+
+  // calculate errors for the training data set
+
+  if ((traindata.training_analysis) && ((int) traindata.data_path.size() > 1) )
+    error_analysis(traindata, desc.c);
+
+  // calculate errors for the test data set
+
+  if ((testdata.test_analysis) && ((int) testdata.data_path.size() > 1) && (testdata.data_path != traindata.data_path))
+    error_analysis(testdata, desc.c);
+
+  // calculate energy and force for the training data set
+
+  if ((traindata.training_calculation) && ((int) traindata.data_path.size() > 1) )
+    energyforce_calculation(traindata, desc.c);
+
+  // calculate energy and force for the test data set
+
+  if ((testdata.test_calculation) && ((int) testdata.data_path.size() > 1) && (testdata.data_path != traindata.data_path) )
+    energyforce_calculation(testdata, desc.c);
+
+  // deallocate training data
+
+  if ((int) traindata.data_path.size() > 1){
+    memory->destroy(traindata.lattice);
+    memory->destroy(traindata.energy);
+    memory->destroy(traindata.stress);
+    memory->destroy(traindata.position);
+    memory->destroy(traindata.force);
+    memory->destroy(traindata.atomtype);
+  }
+
+  // deallocate testing data
+
+  if ((int) testdata.data_path.size() > 1 && (testdata.data_path != traindata.data_path)){
+    memory->destroy(testdata.lattice);
+    memory->destroy(testdata.energy);
+    memory->destroy(testdata.stress);
+    memory->destroy(testdata.position);
+    memory->destroy(testdata.force);
+    memory->destroy(testdata.atomtype);
+  }
+
+  // deallocate descriptors
+
+  memory->destroy(desc.gd);
+  memory->destroy(desc.gdd);
+  memory->destroy(desc.A);
+  memory->destroy(desc.b);
+  memory->destroy(desc.c);
+  memory->destroy(desc.tmpint);
+
+  // deallocate neighbor data
+
+  memory->destroy(nb.alist);
+  memory->destroy(nb.pairnum);
+  memory->destroy(nb.pairnum_cumsum);
+  memory->destroy(nb.pairlist);
+  memory->destroy(nb.y);
+}
+
+/* ---------------------------------------------------------------------- */
+
+void CFITPOD::read_data_file(double *fitting_weights, std::string &file_format,
+                             std::string &file_extension, std::string &test_path,
+                             std::string &training_path, const std::string &data_file)
+{
+  std::string datafilename = data_file;
+  FILE *fpdata;
+  //if (comm->me == 0){
```

Comment:
This commented out code should be reinstated. When running in parallel you do not want every parallel process to open the same file. This can wreak havoc on parallel file system and be worse on non-parallel networked file systems. Rather you want only MPI rank 0 to read it and then broadcast the data to all MPI ranks, or have MPI rank 0 read *and* parse the file and then broadcast the resulting parsed data.